tonic/src/status.rs

```diff
@@ -359,7 +364,13 @@ impl Status {
             _ => Code::Unknown,
         };
 
-        Status::new(code, format!("h2 protocol error: {}", err))
+        let mut status = Self::new(code, format!("h2 protocol error: {}", err));
+        let error: Option<Box<dyn Error + Send + Sync + 'static>> = err
+            .reason()
+            .map(h2::Error::from)
+            .map(|err| Box::new(err) as Box<dyn Error + Send + Sync + 'static>);
```

Comment:
I'm not sure if this could be problematic but if the `h2::Error` doesn't have a `reason` we don't store it as the source. I'm wondering if it would make sense to fix that.

It would require some further since this method would have to receive an owned `h2::Error` which cascades into a few more places in the code. `try_from_error` for example should probably get a `crate::Error` (alias for `Box<dyn Error + Send + Sync + 'static>`). I briefly looked around at the callsites and it seems to be doable.

But again I'm not sure if thats actually necessary. @LucioFranco what do you think?

Reply:
> It would require some further since this method would have to receive an owned h2::Error which cascades into a few more places in the code. try_from_error for example should probably get a crate::Error (alias for Box<dyn Error + Send + Sync + 'static>). I briefly looked around at the callsites and it seems to be doable.

Correct, that is why I didn't just store the `h2::Error` here since it would require `h2::Error` to be `Clone`.

Also, using an owned value for this conversion would probably be better done via a `TryFrom` impl?

Reply:
Yeah, I wonder if we can make it so we get an owned copy of h2::Error and then store it. That would then allow the user to get the h2 error -> source error?

Reply:
That would work when converting `h2::Error` directly by changing `Status::from_h2_error` to take an owned value.

It would not work for `Status::try_from_error` though because the chain of errors retrieved via `.source` are references to the errors. There does not seem to be an API on `Error` to unwrap a chain of errors (e.g., a form of `into_inner`).

Reply:
Hm yeah thats a bit unfortunate. Although I pulled down your branch and found a solution that might work. If we change `Status::try_from_error` to

```rust
impl Status {
    fn try_from_error(err: Box<dyn Error + 'static>) -> Result<Status, Box<dyn Error + 'static>> {
        let err = match err.downcast::<Status>() {
            Ok(status) => {
                return Ok(*status);
            }
            Err(err) => err,
        };

        #[cfg(feature = "transport")]
        let err = match err.downcast::<h2::Error>() {
            Ok(h2) => {
                return Ok(Status::from_h2_error(*h2));
            }
            Err(err) => err,
        };

        if let Some(status) = find_status_in_cause_chain(&*err) {
            return Ok(status);
        }

        Err(err)
    }
}

fn find_status_in_cause_chain(err: &(dyn Error + 'static)) -> Option<Status> {
    let mut cause = Some(err);

    while let Some(err) = cause {
        if let Some(status) = err.downcast_ref::<Status>() {
            return Some(Status {
                code: status.code,
                message: status.message.clone(),
                details: status.details.clone(),
                metadata: status.metadata.clone(),
                // Since `Status` is not `Clone`, any `source` on the original Status
                // cannot be cloned so must remain with the original `Status`.
                source: None,
            });
        }

        cause = err.source();
    }

    None
}
```

That captures almost all the previous behavior. The use case that doesn't work where someone has an `h2::Error` thats nested inside the chain of causes. Is that a common use case, probably not but I don't know. There isn't a test for. It does support `Status` being nested inside the chain, which there is a test for so I figure its something we want to keep working.

Reply:
I've incorporated this change into the PR. Please let me know how it looks. I had to convert call sites to box and move errors.

>  Is that a common use case, probably not but I don't know. 

I don't know either.